sylvia-derive/src/message.rs

```diff
@@ -663,35 +663,62 @@ impl<'a> GlueMessage<'a> {
         let dispatch_arm = quote! {#contract_name :: #contract (msg) =>msg.dispatch(contract, ctx)};
 
         let deserialization_attempts = interfaces.iter().map(|interface| {
-            let ContractMessageAttr { variant, .. } = interface;
-            let var = Ident::new(&variant.to_string().to_case(Case::Snake), variant.span());
-
+            let ContractMessageAttr {
+                module, variant, ..
+            } = interface;
             quote! {
-                let #var = match val.clone().deserialize_into() {
-                    Ok(msg) => return Ok(Self:: #variant (msg)),
-                    Err(err) => err,
+                let msgs = &#module :: #name ::messages();
+                let map = match val.clone() {
+                    #sylvia ::serde_value::Value::Map(map) => map,
+                    _ => panic!("Unexpected msg format!")
+                };
+                if map.len() != 1 {
+                    panic!("Unexpected number of elements received to deserialize!")
+                }
+                let k = match map.into_iter().next() {
+                    Some(k) => k,
+                    None => panic!("Couldn't access the only element of map!")
                 };
+                if let #sylvia ::serde_value::Value::String(k) = k.0 {
+                    if msgs.into_iter().any(|msg| msg == &k) {
+                        match val.clone().deserialize_into() {
+                            Ok(msg) => return Ok(Self:: #variant (msg)),
+                            Err(err) => return Err(D::Error::custom(format!(
+                                    "Deserialization failed for {}\n{}",
+                                    stringify!(k), stringify!(err)
+                                )))
```

Comment:
Here I would literally just forward an error. We do not need to add more context. So here just `val.deserialize_into().map(Self:: #variant)` and be happy.